### ISSUE TITLE
Feature request: always display references

### ISSUE THREAD

**Issue by mattetti** [Microsoft/vscode-go#726]
This is probably a hard one but something that is super useful when writing C# in Visual Studio

The feature is defined here: https://code.visualstudio.com/docs/editor/editingevolved#_reference-information
![](https://code.visualstudio.com/images/editingevolved_referenceinfo.png)

VSCode supports this feature when asking for references of a specific value, but the data isn't cached and it isn't always displayed as in the image above.

https://code.visualstudio.com/Docs/extensionAPI/vscode-api#_a-namecodelensaspan-classcodeitem-id471codelensspan

**Comment by redpixr** [comment]
++ this is like a must have time saver to me.

**Comment by theSoenke** [comment]
This would be really awesome. Codelens is also now available for Typescript https://code.visualstudio.com/updates/v1_9#_typescript-references-codelens It's has been a very useful feature for me in C# in VS

**Comment by ramya-rao-a** [comment]
Yes, it is a very good feature indeed. PRs are welcome.

**Comment by theSoenke** [comment]
I've build a go codelens prototype. It shows references for functions, interfaces and structs. It takes some time to find all references, but i don't think there is much i can do about it. https://github.com/thesoenke/vscode-go/tree/codelens

![go-codelens](https://cloud.githubusercontent.com/assets/1757301/25073820/fd20c324-22ee-11e7-9bae-546e07cc8cdf.png)

One issue i ran into is that go-outline seems to return always 1 as the start character instead as of at least 6 for functions in a line, but for structs it's correct.

Another issue is that the console is filled with errors from guru when editing the code. Not sure what's the best solution here

**Comment by ramya-rao-a** [comment]
@theSoenke great work! Please send a PR and we can try and get this in the next release :)

Few points
- I have a case where a function is used in another Go project outside the workspace opened in VS Code. The Codelens lists this external reference, but when I do a "Find all references", the external reference is not listed. Any idea how that happened? 
- About `go-outline` returning different offset for `func` vs `type`, looks like that is how the Go parser works. See https://github.com/lukehoban/go-outline/blob/master/main.go#L53
- About console filled with errors, codelens tries to update itself while you edit the file. And with every edit, the offset changes, but codelens still looks at the old offsets, thus causing the errors. Well, that's my current theory anyway, am digging deeper to see how other languages have done this and see if I can come up with a solution

**Comment by ramya-rao-a** [comment]
@theSoenke 

About the errors in the console on edit, I see 2 types

- `guru` failing. This I am looking into
- Other one is
<img width="565" alt="screen shot 2017-04-17 at 12 45 26 pm" src="https://cloud.githubusercontent.com/assets/16890566/25102161/076625f4-236c-11e7-94d2-033caf380a4b.png">

The second one is because you are returning the codelens promises even the ones that do not return a codelens.

So something like this
```
let codelenses = [];
            return Promise.all(lenses).then(values => {
                values.forEach(lens => {
                    if (lens){
                        codelenses.push(lens);
                    }
                });
                return codelenses;
            });
```

instead of `return Promise.all(lenses)` will fix that.

**Comment by ramya-rao-a** [comment]
@theSoenke Found the issue. Both `guru` and `go-outline` work only on saved files. Therefore, when you edit, all the positions get messed up. I'll fix that and this should work like a charm.

**Comment by ramya-rao-a** [comment]
@theSoenke  Alright, I have updated both `go-outline` and `guru` to use the `-modified` flag to work on unsaved file contents.

Merge from master, and things should work well now.

**Comment by theSoenke** [comment]
Great! Thanks for looking so quick into this and provide the fixes :wave: 

Unfortunately the debug console still contains errors from guru when syntax is not valid and the file not saved.

**Comment by ramya-rao-a** [comment]
Oh that! That is the call to references being cancelled.

Change the `if (err)` at https://github.com/theSoenke/vscode-go/blob/codelens/src/goReferences.ts#L44 to `if (err && (<any>err).killed !== true`, and you should be good to go

**Comment by ramya-rao-a** [comment]
Or check the cancellation token in `goReferences` before running `guru`

**Comment by ramya-rao-a** [comment]
This feature is now available in the latest update (0.6.60).
It can be disabled by setting `go.referencesCodeLens.enable` to `false`

**Comment by theSoenke** [comment]
Happy to contribute a little bit to this great extension :)
